buffer.go

```diff
@@ -0,0 +1,159 @@
+package stats
+
+import (
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+)
+
+// Buffer is the implementation of a measure handler which uses a Serializer to
+// serialize the metric into a memory buffer and write them once the buffer has
+// reached a target size.
+type Buffer struct {
+	// Target size of the memory buffer where metrics are serialized.
+	//
+	// If left to zero, a size of 1024 bytes is used as default (this is low,
+	// you should set this value).
+	BufferSize int
+
+	// The Serializer used to write the measures.
+	//
+	// This field cannot be nil.
+	Serializer Serializer
+
+	buffer unsafe.Pointer
```

Comment:
The intent here is to have an algorithm that is mostly wait-free, this gives throughput that increases linearly with the number of cores. If we were using a single buffer (circular or not) then goroutines need to wait for flush operations to complete when the buffer is full. We need to use a dynamic number of buffers for maximum throughput (this is what the sync.Pool gives us) but we also need to avoid lock contention on transition phases, like when replacing a full buffer. Using atomic compare-and-swap we can ensure that no goroutine is contending on a lock, when they do conflicting things like having two goroutines allocate a new buffer only one will succeed and the buffer will give its buffer back to the pool.